quic/s2n-quic-core/src/crypto/tls/offload.rs

```diff
@@ -0,0 +1,567 @@
+                Poll::Pending => break,
+            };
+
+            match req {
+                Request::HandshakeKeys(key, header_key) => {
+                    context.on_handshake_keys(key, header_key)?;
+                }
+                Request::ZeroRttKeys(key, header_key, transport_parameters) => {
+                    context.on_zero_rtt_keys(
+                        key,
+                        header_key,
+                        tls::ApplicationParameters {
+                            transport_parameters: &transport_parameters,
+                        },
+                    )?;
+                }
+                Request::ClientParams(client_params, mut server_params) => context
+                    .on_client_application_params(
+                        tls::ApplicationParameters {
+                            transport_parameters: &client_params,
+                        },
+                        &mut server_params,
+                    )?,
+                Request::OneRttKeys(key, header_key, transport_parameters) => {
+                    context.on_one_rtt_keys(
+                        key,
+                        header_key,
+                        tls::ApplicationParameters {
+                            transport_parameters: &transport_parameters,
+                        },
+                    )?;
+                }
+                Request::Done(session, res) => {
+                    self.inner = Some(session);
+                    self.is_poll_done = Some(res);
+
+                    return Poll::Ready(res);
+                }
+                Request::ServerName(server_name) => {
+                    context.on_server_name(server_name)?;
+                }
+                Request::ApplicationProtocol(application_protocol) => {
+                    context.on_application_protocol(application_protocol)?;
+                }
+                Request::HandshakeComplete => {
+                    context.on_handshake_complete()?;
+                }
+                Request::CanSendInitial(sender) => {
+                    let _ = sender.send(context.can_send_initial());
+                }
+                Request::ReceiveInitial(max_len, sender) => {
+                    let resp = context.receive_initial(max_len);
+                    let _ = sender.send(resp);
+                }
+                Request::ReceiveApplication(max_len, sender) => {
+                    let resp = context.receive_application(max_len);
+                    let _ = sender.send(resp);
+                }
+                Request::ReceiveHandshake(max_len, sender) => {
+                    let resp = context.receive_handshake(max_len);
+                    if resp.is_some() {
+                        // We need to wake up the s2n-quic endpoint after providing
+                        // handshake packets to the TLS provider as there may now be
+                        // handshake data that needs to be sent in response.
```

Comment:
what is special about the handshake packets that require this? ie why doesn't `ReceiveInitial` need it too?